File: Programmers/Level2/괄호변환/괄호변환.py
def divide(p):
    p1, p2 = 0, 0
    
    for i in range(len(p)):
        if p[i] == '(':
            p1 += 1
        else:
            p2 += 1
        
        if p1 == p2:
            return p[:p1 + p2], p[p1 + p2:]

def check(u):
    stack = []
    for i in u:
        if i == '(':
            stack.append(i)
        else:
            if not stack:
                return False
            stack.pop()
        return True
    
def solution(p):
    if not p:
        return ''
    
    u, v = divide(p)
    
    if check(u):
        return u + solution(v)
    else:
        answer = '(' + solution(v) + ')'
        
        for i in u[1:len(u) - 1]:
            answer += ')' if i == '(' else '('
    
    return answer

File: Programmers/Level2/괄호변환/괄호변환.js
function divide(w) {
  let w1 = 0,
    w2 = 0;
  [];
  for (let i = 0; i < w.length; i++) {
    w[i] == '(' ? w1++ : w2++;
    if (w1 === w2) return [w.slice(0, w1 + w2), w.slice(w1 + w2)];
  }
}

function check(u) {
  const stack = [];
  for (let chr of u) {
    if (chr === '(') {
      stack.push(chr);
    } else {
      if (!stack.length) return false;
      stack.pop();
    }
    return true;
  }
}

function solution(p) {
  if (!p.length) return '';

  let [u, v] = divide(p);

  if (check(u)) return u + solution(v);
  else {
    const tmp = [...u.slice(1, -1)]
      .map(item => (item === '(' ? ')' : '('))
      .join('');
    return '(' + solution(v) + ')' + tmp;
  }
}
